plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp

```diff
@@ -2449,6 +2468,12 @@ double CudaCalcAmoebaVdwForceKernel::execute(ContextImpl& context, bool includeF
         hasInitializedNonbonded = true;
         nonbonded->initialize(system);
     }
+
+    // Not sure about passing a Context parameter to the Kernal?
+    vector<float> vdwLambdaVec(1, 0);
+    vdwLambdaVec[0] = context.getParameter(AmoebaVdwForce::Lambda());
+    vdwLambda.upload(vdwLambdaVec);
```

Comment:
This will work, but it's a little inefficient.  The problem is that `upload()` is a blocking call.  So it doesn't return until the upload is complete, and at that point the GPU's work queue will have been emptied out.  Then it will take some time for it to set up the next kernel, and during that time the GPU won't be doing any work.  It's not that big a deal, but still a slight inefficiency.

The first thing you can do is just put it in an "if" block based on the alchemical method.  In most simulations it will be `None`, so you can just skip this altogether.

The next thing you can do is use the alternate version of `upload()` (https://github.com/openmm/openmm/blob/master/platforms/cuda/include/CudaArray.h#L176) that supports non-blocking copies.  Just pass `false` for the second argument.  But of course there's a complication: non-blocking copies can only be done from pinned memory.  So to do that, you need to allocate the buffer with `cuMemHostAlloc()`, then free it with `cuMemFreeHost()` in the destructor.  For an example of this, take a look at the `pinnedBuffer` field of CudaContext.

Reply:
Here's another way of optimizing it: just keep track of what the last value was that you uploaded, and only upload a new value when it changes.  Lambda won't usually change every time step.  If you're only having to upload a new value occasionally, the cost will be totally negligible.